File: Swift/MyLisp/MyLisp/interpreter.swift
//
//  interpreter.swift
//  MyLisp
//
//  Created by 赵雅琦 on 2020/4/8.
//  Copyright © 2020 z. All rights reserved.
//

import Foundation

//func caculate(ast:Array<Any>, env:Env) -> Any {
fileprivate func isNumber(_ ast: Any) -> Bool {
    return ast is Int || ast is Float
}

fileprivate func isSymbol(_ ast: Any) -> Bool {
    return ast is String
}

fileprivate func isArray(_ ast: Any) -> Bool {
    return ast is Array<Any>
}

//func caculate(ast:Any, env:Env) throws -> Any {
//func caculate(ast:Any, env:Env) throws -> Any {
func caculate(ast:Any, env:Env) throws -> Any {
    print(ast)
    let env = Env(env: env)
    if isNumber(ast) {
        return ast
    }
    else if isSymbol(ast) {
        return try env.lookup(key: ast as! String)
    }
    else if isArray(ast) {
//        let astArray = ast as! Array<Any>
        let ast = ast as! Array<Any>
        if ast.count == 2 {
            let ast1 = ast[0]
            let ast2 = ast[1]
//            let (ast1, ast2) = ast;
            if ast1 is String {             // call, (exp1 exp)
                let exp = try env.lookup(key: ast1 as! String)
                return try caculate(ast: [exp ?? nil, ast2], env: env)
            }
            else if ast1 is Array<Any> && (ast1 as! Array<Any>).count >= 3 {
                let key = (ast1 as! Array<Any>)[0] as? String
                if key == "lambda" || key == "λ" {
                    let ast = ast1 as! Array<Any>
                    let params = ast[1] as! Array<Any>
                    let param1 = params[0]
//                    env.extend(key: param1 as! String, value: ast2)
                    let expr = try caculate(ast: ast2, env: env)
                    try env.extend(key: param1 as! String, value: expr)
//                    return try caculate(ast: (ast2 as! Array<Any>)[2], env: env)
                    return try caculate(ast: ast[2], env: env)
                }
                else {
                    throw EvaluationError.Three
                }
            }
            else {
                throw EvaluationError.Two
            }
        }
        else if ast.count == 3 {
            let exp1 = ast[0] as! String
            let exp2 = ast[1]
            let exp3 = ast[2]
//                        let [exp1,exp2,exp3] = ast
//            let op = exp1 as! String
            if exp1 == "let" {
                let bindings = exp2 as! Array<Any>
                if bindings.count > 0 {
                    for binding in bindings {
                        let binding = binding as! Array<Any>
                        if binding.count == 2 {
                            try env.extend(key: binding[0] as! String, value: binding[1])
                        }
                        else {
                            throw EvaluationError.BindingError2
                        }
                        return try caculate(ast: exp3, env: env)
                    }
                }
                else {
                    throw EvaluationError.BindingError1
                }
            }
            else {
                let x = try caculate(ast: exp2, env: env) as! Int
                let y = try caculate(ast: exp3, env: env) as! Int
                var result = 0
                switch exp1 {
                case "+": result = x + y
                case "-": result = x - y
                case "*": result = x * y
                case "/": result = x / y
                default: throw EvaluationError.OpNotSupported
                }
                return result
            }
        }
        else {
            throw EvaluationError.One
        }
    }
    else {
        throw EvaluationError.Zero
    }
     
    return ""
}

File: MyLisp/src/main/kotlin/Interpreter.kt

fun calc(ast: Any, env: Env): Double {
//    var env = env.copy()
    var env = Env(env)

    smartPrint("ast: $ast")
    if (ast == null) {
        throw Exception("Error: ast is null!")
    } else {
        when {
//            ast is Number -> {
            ast is kotlin.Number -> {
                smartPrint("ast:$ast is a Number")
                return ast.toDouble()
            }
            ast is String -> {
//                return env.lookup(ast) as Double
                return (env.lookup(ast) as Integer).toDouble()
            }
            (ast is List<*>).not() -> {
                smartPrint("ast:$ast is not an Array")
                error("Error: not a expression!")
            }
            else -> {
                smartPrint("ast:$ast is an Array")
                var op = (ast as List<Any>)[0]

                smartPrint("op is $op")
//                var v1 = calc(ast[1], env)
//                var ast0 = ast[0]
                var ast1 = ast[1]
//                var ast2: Any? = null
                var ast2: Any = 0
//                    calc(ast[2], env)
                //                    v2 = calc(ast[2] as List<*>, env)
                //                    v2 = ast[2]?.let { calc(it, env) }
                if (ast.count() > 2) ast2 = ast[2]

//                var fun1 = { calc(ast1, env)}
//                var fun2 = {
//                    if (ast2 != null) {
//                        calc(ast2, env)
//                    }
//                    else {
//                        0
//                    }
//                }

                smartPrint("numbers is $ast1 and $ast2")

                when (op) {
                    is List<*> -> {         //((lambda (x) exp) exp)
//                        var keyword = (ast1 as List<*>)[0]
                        var keyword = (op as List<*>)[0]
                        if (op.count() >= 3 && (keyword == "lambda" || keyword == "λ")) {
//                            var params = ast1[1] as List<*>
                            var params = op[1] as List<*>
                            var param1 = params[0] as String
                            env.extend(param1, ast1)
//                            return ast1[2]?.let { calc(it, env) }!!
                            return calc(op[2]!!, env)
                        } else {
                            error("error: Unsupported operation!")
                        }
                    }
                    "let" -> {
                        var bindings = ast1 as List<*>
                        if (bindings == null || bindings.count() == 0) {
                            error("let clause error: bindings error 1!")
                        } else {
                            var binding = bindings[0] as List<*>
//                            binding[1]?.let { env.extend(binding[0] as String, it) }!!
//                            binding[1]?.let { env.extend(binding[0] as String, it) }!!
                            env.extend(binding[0] as String, binding[1]!!)
                            return calc(ast2, env)
                        }
                    }
                    "+" -> {
//                        return ast1 + ast2
//                        return fun1() + fun2()
                        return calc(ast1, env) + calc(ast2, env)
                    }
                    "-" -> {
//                        return ast1 - ast2
                        return calc(ast1, env) - calc(ast2, env)
                    }
                    "*" -> {
//                        return ast1 * ast2
                        return calc(ast1, env) * calc(ast2, env)
                    }
                    "/" -> {
//                        return ast1 / ast2
                        return calc(ast1, env) / calc(ast2, env)
                    }
                    else -> {   // call, (exp1 exp)
//                        var expAst = env.lookup(ast1 as String)
                        var expAst = env.lookup(op as String)
                        smartPrint(expAst)
//                        return calc(listOf(expAst, ast2), env)
                        return calc(listOf(expAst, ast1), env)
                    }
                }
            }
        }
    }
}

fun calc(ast: Any): Double {
    return calc(ast, Env())
}